src/libvirt-xml-update.js

```diff
@@ -296,3 +296,56 @@ export function updateMaxMemory(domXml, maxMemory) {
 
     return s.serializeToString(doc);
 }
+
+function xmlGetXPath(doc, xpath) {
+    const elements = [];
+    let node = null;
+    const iterator = doc.evaluate(xpath, doc);
+    while ((node = iterator.iterateNext()))
+        elements.push(node);
+    return elements;
+}
+
+export function replaceSpice(domXml) {
+    // this implements https://access.redhat.com/solutions/6955095
+    logDebug("replaceSpice original XML:", domXml);
+
+    const doc = getDoc(domXml);
+
+    xmlGetXPath(doc, "/domain/devices/audio[@type='spice']").forEach(n => n.remove());
+    xmlGetXPath(doc, "//domain/devices/redirdev[@type='spicevmc']").forEach(n => n.remove());
+    xmlGetXPath(doc, "//domain/devices/channel[@type='spicevmc']").forEach(n => n.remove());
+
+    xmlGetXPath(doc, "//domain/devices/video").forEach(video => {
+        const model = getSingleOptionalElem(video, "model");
+        if (model?.getAttribute("type") === "qxl") {
+            // remove all attributes except for primary; don't change list while iterating
+            const removeAttrs = [];
+            for (let i = 0; i < model.attributes.length; i++) {
+                const attr = model.attributes.item(i);
+                if (attr.name !== "primary")
+                    removeAttrs.push(attr.name);
```

Comment:
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1387-20240119-062100-f9c8ea53-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)